File: docs/engine-evaluation/run_zen.py
"""Runs the same zen-policy.json through the Python binding (what a Python 3.12 Lambda would use)."""
import json
import sys
import time

sys.path.insert(0, "pylib")
import zen  # noqa: E402

with open("zen-policy.json", "rb") as fh:
    content = fh.read()

engine = zen.ZenEngine()
decision = engine.create_decision(content.decode("utf-8"))

for label, path in (("scenarios", "cases.json"), ("generated", "fuzz.json")):
    with open(path, encoding="utf-8") as fh:
        cases = json.load(fh)
    match = 0
    diffs = []
    t0 = time.perf_counter()
    for c in cases:
        r = decision.evaluate(c["input"])["result"]
        got = {
            "decision": r["decision"],
            "hard": sorted(set(r["hard"])),
            "soft": sorted(set(r["soft"])),
            "score": r["score"],
        }
        e = c["expected"]
        if got["decision"] == e["decision"] and got["hard"] == e["hard"] and got["soft"] == e["soft"] and got["score"] == e["score"]:
            match += 1
        elif len(diffs) < 3:
            diffs.append({"id": c["id"], "exp": e, "got": got})
    ms = (time.perf_counter() - t0) * 1000 / len(cases)
    print(f"zen-python/{label}: {match}/{len(cases)} match, {ms:.3f} ms per evaluation")
    for d in diffs:
        print("  DIFF", d)

File: docs/engine-evaluation/run-zen.mjs
// Runs cases.json + fuzz.json through GoRules Zen and compares with the reference engine.
import { readFileSync } from "node:fs";
import { ZenEngine } from "@gorules/zen-engine";

const content = readFileSync("zen-policy.json");
const engine = new ZenEngine();
const decision = engine.createDecision(content);

const sets = { scenarios: JSON.parse(readFileSync("cases.json", "utf8")), generated: JSON.parse(readFileSync("fuzz.json", "utf8")) };
const eq = (a, b) => JSON.stringify(a) === JSON.stringify(b);

for (const [label, cases] of Object.entries(sets)) {
  let match = 0;
  const diffs = [];
  const t0 = performance.now();
  for (const c of cases) {
    const { result } = await decision.evaluate(c.input);
    const got = { decision: result.decision, hard: [...new Set(result.hard)].sort(), soft: [...new Set(result.soft)].sort(), score: result.score };
    const exp = c.expected;
    if (got.decision === exp.decision && eq(got.hard, exp.hard) && eq(got.soft, exp.soft) && got.score === exp.score) match++;
    else if (diffs.length < 5) diffs.push({ id: c.id, exp, got });
  }
  const ms = performance.now() - t0;
  console.log(`zen/${label}: ${match}/${cases.length} match, ${(ms / cases.length).toFixed(3)} ms per evaluation`);
  for (const d of diffs) console.log("  DIFF", JSON.stringify(d));
}
engine.dispose?.();
